cpp/src/text/replace.cu

```diff
@@ -0,0 +1,172 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/strings/detail/utilities.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/utilities/error.hpp>
+#include <nvtext/detail/tokenize.hpp>
+#include <nvtext/tokenize.hpp>
+#include <strings/utilities.cuh>
+#include <text/utilities/tokenize_ops.cuh>
+
+namespace nvtext {
+namespace detail {
+namespace {
+
+using strings_iterator = cudf::column_device_view::const_iterator<cudf::string_view>;
+
+/**
+ * @brief Functor to replace tokens in each string.
+ *
+ * This tokenizes a string using the given d_delimiter and replaces any tokens that match
+ * a string in d_targets_begin/end with those from the d_repls column.
+ * Strings with no matching tokens are left unchanged.
+ *
+ * This should be called first to compute the size of each output string and then a second
+ * time to fill in the allocated output buffer for each string.
+ */
+struct replace_tokens_fn {
+  cudf::column_device_view const d_strings;  ///< strings to tokenize
+  strings_iterator d_targets_begin;          ///< strings to search for
+  strings_iterator d_targets_end;
+  cudf::column_device_view const d_repls;  ///< replacement strings
+  cudf::string_view const d_delimiter;     ///< delimiter characters for tokenizing
+  const int32_t* d_offsets{};              ///< for locating output string in d_chars
+  char* d_chars{};                         ///< output buffer
+
+  __device__ cudf::size_type operator()(cudf::size_type idx)
+  {
+    if (d_strings.is_null(idx)) return 0;
+    auto const d_str = d_strings.element<cudf::string_view>(idx);
+    auto nbytes      = d_str.size_bytes();
+    auto in_ptr      = d_str.data();
+    auto out_ptr     = d_chars ? d_chars + d_offsets[idx] : nullptr;
+    auto last_pos    = cudf::size_type{0};
+    characters_tokenizer tokenizer(d_str, d_delimiter);
+    while (tokenizer.next_token()) {
+      auto token_pos = tokenizer.token_byte_positions();
+      cudf::string_view token{d_str.data() + token_pos.first, (token_pos.second - token_pos.first)};
+      // check if the token matches any of the targets
+      strings_iterator found_itr = thrust::find(thrust::seq, d_targets_begin, d_targets_end, token);
+      if (found_itr != d_targets_end) {  // match found
+        // retrieve the corresponding replacement string or
+        // if only one repl string, use that one for all targets
+        cudf::size_type repl_idx = thrust::distance(d_targets_begin, found_itr);
+        cudf::string_view d_repl = d_repls.size() == 1
+                                     ? d_repls.element<cudf::string_view>(0)
+                                     : d_repls.element<cudf::string_view>(repl_idx);
```

Comment:
I was working on trying to get this working locally bust messed up my branch, will come back to tomorrow. Potentially the lambda could be a function called `replacement_string`
```cpp
    if (d_strings.is_null(idx)) return 0;

    auto const d_str = d_strings.element<cudf::string_view>(idx);
    auto nbytes      = d_str.size_bytes();
    auto in_ptr      = d_str.data();
    auto out_ptr     = d_chars ? d_chars + d_offsets[idx] : nullptr;
    auto last_pos    = cudf::size_type{0};
    auto tokenizer   = characters_tokenizer{d_str, d_delimiter};

    while (tokenizer.next_token()) {
      auto const token_pos = tokenizer.token_byte_positions();
      auto const token_len = token_pos.second - token_pos.first;
      auto const token     = cudf::string_view{d_str.data() + token_pos.first, token_len};
      auto const found_itr = thrust::find(thrust::seq, d_targets_begin, d_targets_end, token);

      if (found_itr != d_targets_end) {  // match found

        auto const d_repl = [&] {
          // retrieve the corresponding replacement string or
          // if only one repl string, use that one for all targets
          auto const repl_idx = thrust::distance(d_targets_begin, found_itr);
          return d_repls.size() == 1 ? d_repls.element<cudf::string_view>(0)
                                     : d_repls.element<cudf::string_view>(repl_idx);
        }();
```